### PR TITLE
8335912: Add an operation mode to the jar command when extracting to not overwriting existing files

### PR DESCRIPTION
**Pull request by slowhog**
This PR support a -k, --keep options like tar that allows jar to avoid override existing files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8340685](https://bugs.openjdk.org/browse/JDK-8340685) to be approved

### Issues
 * [JDK-8335912](https://bugs.openjdk.org/browse/JDK-8335912): Add an operation mode to the jar command when extracting to not overwriting existing files (**Enhancement** - P5)
 * [JDK-8340685](https://bugs.openjdk.org/browse/JDK-8340685): Add -k, --keep-old-files to the jar command when extracting to not overwriting existing files (**CSR**)

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21141/head:pull/21141` \
`$ git checkout pull/21141`

Update a local copy of the PR: \
`$ git checkout pull/21141` \
`$ git pull https://git.openjdk.org/jdk.git pull/21141/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21141`

View PR using the GUI difftool: \
`$ git pr show -t 21141`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21141.diff">https://git.openjdk.org/jdk/pull/21141.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21141#issuecomment-2368939102)

### CODE REVIEW

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties [hunk 1]
```diff
@@ -307,6 +309,12 @@ main.help.opt.create.update.index.date=\
 \      --date=TIMESTAMP       The timestamp in ISO-8601 extended offset date-time with\n\
 \                             optional time-zone format, to use for the timestamps of\n\
 \                             entries, e.g. "2022-02-12T12:30:00-05:00"
+main.help.opt.extract=\
+\ Operation modifiers valid only in extract mode:\n
+main.help.opt.extract.keep-old-files=\
+\  -k, --keep-old-files       Do not overwrite existing files.\n\
+\                             In particular, if a file appears more than once in an\n\
```
Comment (LanceAndersen) [inline src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:316]: In addition, we need to update the help to clarify -x. --extract behavior that it will overwrite by default.

Here is the verbiage from tar to give you a start

` -x      Extract to disk from the archive.  If a file with the same name appears more than
             once in the archive, each copy will be extracted, with later copies overwriting
             (replacing) earlier copies.  The long option form is --extract.`

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties [hunk 2]
```diff
@@ -165,6 +165,8 @@ out.create=\
         \ \ created: {0}
 out.extracted=\
         extracted: {0}
+out.kept=\
+        \ \ skipped: {0}
```
Comment (LanceAndersen) [inline src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:169]: We might want to add a bit more wording to indicate it is being skipped due to the file already existing on disk

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    @AfterMethod
+    public void run() {
+        if (onCompletion != null) {
+            onCompletion.run();
+        }
+    }
+
+    @Test
+    public void testExtract() throws IOException {
```
Comment (LanceAndersen) [inline test/jdk/tools/jar/ExtractFilesTest.java:94]: Please consider adding comments introducing the methods used by the test

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+/*
+ * @test
+ * @bug 8335912
+ * @summary test extract jar files overwrite existing files behavior
+ * @library /test/lib
+ * @modules jdk.jartool
+ * @build jdk.test.lib.Platform
+ *        jdk.test.lib.util.FileUtils
+ * @run testng ExtractFilesTest
```
Comment (LanceAndersen) [inline test/jdk/tools/jar/ExtractFilesTest.java:32]: Please convert the test to use junit as we are moving away from testng

**Review by LanceAndersen — COMMENTED** [review]
Thank you for tackling Henry.

A few initial comments.  Also please make sure to update the copyright on the files being updated to 2024

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 1 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:319]: Updated.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 2 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:169]: I follow existing pattern with short status update. Open to suggestions.

File: test/jdk/tools/jar/ExtractFilesTest.java (hunk 3 above)
Comment (slowhog) [reply test/jdk/tools/jar/ExtractFilesTest.java:85]: Updated.

File: test/jdk/tools/jar/ExtractFilesTest.java (hunk 4 above)
Comment (slowhog) [reply test/jdk/tools/jar/ExtractFilesTest.java]: Done. Also add a test case for multiple manifest files in the same archive as we mentioned such use case in the help text.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 1 above)
Comment (jaikiran) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:319]: Hello Henry, I think this `-k` option help text would need a slight modification. Right now it states that if a file appears more than once in an archive, then this setting this flag will not overwrite the earlier copies. In reality, it doesn't matter how many times the file appears in the archive - even if it appears just once, and if the file is being extracted into a directory which already has the same named file at that path, then this `-k` flag will not overwrite that existing file.

So I think we should reword it to talk about its behaviour in context of some file already existing in the destination directory where the jar contents are being extracted. 

I wonder if we should make a mention that we don't do any case sensitive checks for the existence of the file in the destination directory and instead we use filesystem API to check for existence (which depending on the filesystem may be case insensitive, like macosx).

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                " inflated: testfile1" + nl +
+                "  skipped: testfile2" + nl;
+        Assertions.assertEquals("testfile1", cat("testfile1"));
+        Assertions.assertEquals("", cat("testfile2"));
+        rm("META-INF testfile1 testfile2");
+        Assertions.assertArrayEquals(baos.toByteArray(), output.getBytes());
+    }
+
+    private Stream<Path> mkpath(String... args) {
+        return Arrays.stream(args).map(d -> Paths.get(".", d.split("/")));
```
Comment (jaikiran) [inline test/jdk/tools/jar/ExtractFilesTest.java:168]: For newer code usages, like this one, it's recommended to use `Path.of(...)` instead of `Paths.get(...)`.

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        } catch (IOException x) {
+            throw new UncheckedIOException(x);
+        }
+    }
+
+    private String cat(String path) {
+        System.out.println("cat " + path);
+        try {
+            var p = Paths.get(".", path.split("/"));
+            return String.join(nl, Files.readAllLines(p));
```
Comment (jaikiran) [inline test/jdk/tools/jar/ExtractFilesTest.java:207]: Do you think this could be simplified to `return Files.readString(p)`?

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    private void jar(String cmdline) throws IOException {
+        System.out.println("jar " + cmdline);
+        baos.reset();
+
+        // the run method catches IOExceptions, we need to expose them
+        ByteArrayOutputStream baes = new ByteArrayOutputStream();
+        PrintStream err = new PrintStream(baes);
+        PrintStream saveErr = System.err;
+        System.setErr(err);
```
Comment (jaikiran) [inline test/jdk/tools/jar/ExtractFilesTest.java:236]: Given that we are updating the runtime level state, I think we should use `/othervm` in this test's definition to avoid any interference with other parts of the runtime.

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        int rc = JAR_TOOL.run(out, err, cmdline.split(" +"));
+        System.setErr(saveErr);
```
Comment (jaikiran) [inline test/jdk/tools/jar/ExtractFilesTest.java:238]: Doing a
```java
try {
int rc = JAR_TOOL.run(out, err, cmdline.split(" +"));
}finally {
System.setErr(saveErr);
}
```
would be safer.

File: test/jdk/tools/jar/ExtractFilesTest.java [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        if (rc != 0) {
+            String s = baes.toString();
+            if (s.startsWith("java.util.zip.ZipException: duplicate entry: ")) {
```
Comment (jaikiran) [inline test/jdk/tools/jar/ExtractFilesTest.java:241]: This method runs any arbitrary `jar` "command". Is there any significance why we are checking for a "duplicate entry" in the exception message? Maybe we could just remove this entire if block and just throw a `new IOException(s)`? As far as I can see in this test, we don't do any exception type checks on the thrown exception from this method.

File: test/jdk/tools/jar/MultipleManifestTest.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+import java.util.spi.ToolProvider;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipOutputStream;
+
+import jdk.test.lib.util.FileUtils;
+
+@TestInstance(Lifecycle.PER_CLASS)
+@TestMethodOrder(MethodName.class)
```
Comment (jaikiran) [inline test/jdk/tools/jar/MultipleManifestTest.java:67]: Is the use of these annotations intentional? Especially the method ordering?

File: src/jdk.jartool/share/classes/sun/tools/jar/Main.java [hunk 11]
```diff
@@ -582,6 +583,9 @@ boolean parseArgs(String[] args) {
                         case '0':
                             flag0 = true;
                             break;
+                        case 'k':
+                            kflag = true;
```
Comment (LanceAndersen) [inline src/jdk.jartool/share/classes/sun/tools/jar/Main.java:587]: I am wondering if a warning should be displayed if this options is specified with any option other than 'x'

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 2 above)
Comment (LanceAndersen) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:169]: Perhaps something like:

` \ \  file: {0}  exists, skipped`

File: src/jdk.jartool/share/classes/sun/tools/jar/Main.java (hunk 11 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/Main.java:587]: I considered that, but didn't implement it after confirmed other similar options didn't display any warning.
I do think it make sense to show a warning when using an option not valid in specific mode.

File: test/jdk/tools/jar/MultipleManifestTest.java (hunk 10 above)
Comment (slowhog) [reply test/jdk/tools/jar/MultipleManifestTest.java:67]: I had some test failure earlier so I added those to ensure not the race condition or left over even though I don't think that's the case.
I later figured out the real cause, so I think I can remove those.

File: test/jdk/tools/jar/ExtractFilesTest.java (hunk 9 above)
Comment (slowhog) [reply test/jdk/tools/jar/ExtractFilesTest.java:241]: I copied the method from another test that do the create jar, so it does that. Some other comments also related to copied code, I can do a round of clean up.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 1 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:319]: You have valid point. I refer to the tar man page, and I thought do not overwrite existing files says it. The in particular is just to call out a specific use case.

**Comment by eirbjo** [comment]
Is `compiler` label needed? The `jar` tool belongs in `core-libs`, right? This PR does not seem to touch anything in the compiler area?

File: src/jdk.jartool/share/classes/sun/tools/jar/Main.java (hunk 11 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/Main.java:587]: OK, I will add a message. The other option I am talking about is specific mode options, like compression or manifest instead of main options.

**Comment by slowhog** [comment]
> Is `compiler` label needed? The `jar` tool belongs in `core-libs`, right? This PR does not seem to touch anything in the compiler area?

Automatically added by the bot. I agree it's not related to compiler at all. I'll manually remove it.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 2 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:169]: Use this for now to be more close to current style,
`        \ \ skipped: {0} exists`
Was thinking perhaps a different line for detail message, but this should do, as we only skip with file exists.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties (hunk 1 above)
Comment (slowhog) [reply src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties]: Please review the update and see if that's good enough.

File: src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties [hunk 12]
```diff
@@ -316,8 +318,11 @@ main.help.opt.extract=\
 \ Operation modifiers valid only in extract mode:\n
 main.help.opt.extract.keep-old-files=\
 \  -k, --keep-old-files       Do not overwrite existing files.\n\
-\                             In particular, if a file appears more than once in an\n\
-\                             archive, later copies will not overwrite earlier copies.
+\                             If a file with the same name exists in the target\n\
+\                             directory, skip the extraction of that file.\n\
```
Comment (LanceAndersen) [inline src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties:322]: This is OK, but I would probably just  the if the file exists, it will not be overwritten

**Review by LanceAndersen — APPROVED** [review]
Overall,  this looks ok

We will also need an RN which would be good to have available when the CSR is finalized to review

**Review by LanceAndersen — APPROVED** [review]
